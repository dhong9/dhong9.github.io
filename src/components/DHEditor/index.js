// React
import { useState, useEffect } from "react";

// Sections components
import MKInput from "components/MKInput";

// Editor
import { EditorState } from "draft-js";
import { Editor } from "react-draft-wysiwyg";
import { convertToHTML } from "draft-convert";
import DOMPurify from "dompurify";

// CSS
import "react-draft-wysiwyg/dist/react-draft-wysiwyg.css";
import "components/DHEditor/DHEditor.css";

// prop-types is a library for typechecking of props
import PropTypes from "prop-types";

// https://blog.logrocket.com/build-rich-text-editors-react-draft-js-react-draft-wysiwyg/
export default function DHEditor(props) {
  const { isPlainText } = props;

  const [editorState, setEditorState] = useState(() => EditorState.createEmpty());
  const [convertedContent, setConvertedContent] = useState(null);
  const [rootComment, setRootComment] = useState("");

  useEffect(() => {
    const html = convertToHTML(editorState.getCurrentContent());
    setConvertedContent(html);
  }, [editorState]);

  // Sanitizes HTML in editor
  const createMarkup = (html) => ({
    __html: DOMPurify.sanitize(html),
  });

  return isPlainText ? (
    <>
      <Editor
        editorState={editorState}
        onEditorStateChange={setEditorState}
        wrapperClassName="wrapper-class"
        editorClassName="editor-class"
        toolbarClassName="toolbar-class"
      />
      <div className="preview" dangerouslySetInnerHTML={createMarkup(convertedContent)}></div>
    </>
  ) : (
    <MKInput
      variant="standard"
      InputLabelProps={{ shrink: true }}
      multiline
      fullWidth
      rows={6}
      onChange={(e) => setRootComment(e.target.value)}
      value={rootComment}
    />
  );
}

// Typechecking props of FeaturedPost
DHEditor.propTypes = {
  isPlainText: PropTypes.bool,
};

DHEditor.defaultProps = {
  isPlainText: false,
};
