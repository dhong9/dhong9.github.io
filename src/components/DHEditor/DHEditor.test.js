// React testing libraries
import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

// Material Kit 2 React themes
import { ThemeProvider } from "@mui/material/styles";
import theme from "assets/theme";

// Component to test
import DHEditor from "components/DHEditor";

// Mocks
jest.mock("draft-convert", () => {
  return {
    convertFromHTML: jest.fn(),
    convertToRaw: jest.fn(),
    convertToHTML: jest.fn().mockReturnValue("<p>Test content</p>"), // Mock the return value of convertToHTML
  };
});

describe("DHEditor", () => {
  it("renders", () => {
    const ref = React.createRef();

    // By default, isPlainText is false
    const { container } = render(
      <ThemeProvider theme={theme}>
        <DHEditor ref={ref} />
      </ThemeProvider>
    );

    // Get current text
    let val = ref.current.getRootComment();
    expect(val).toEqual("<p>Test content</p>");

    // Set isPlainText to true
    ref.current.handleSetPlainText(true);

    expect(container).toMatchSnapshot();
  });
});
