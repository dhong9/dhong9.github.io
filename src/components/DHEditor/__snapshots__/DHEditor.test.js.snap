// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DHEditor renders 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1c3j9he-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-inw9yj-MuiInputBase-root-MuiInput-root"
    >
      <textarea
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-12marsf-MuiInputBase-input-MuiInput-input"
        id=":r0:"
        rows="6"
        style="height: 0px; overflow: hidden;"
      />
      <textarea
        aria-hidden="true"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-12marsf-MuiInputBase-input-MuiInput-input"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
